terraform/modules/sqs/sqs.tf

```diff
@@ -0,0 +1,53 @@
+variable "environment" { type = "string" }
+
+variable "role" { type = "string" }
+
+variable "account_number" { type = "string" }
+
+variable "region" { type = "string" }
+
+variable "queue_names" {
+  description = "Squad queues"
+  type        = "list"
+  default     = ["celery", "ci_fetch", "ci_poll", "ci_quick", "core_notification", "core_postprocess", "core_quick", "core_reporting"]
+}
+
+resource "aws_sqs_queue" "qa_reports_queue" {
+  count = "${length(var.queue_names)}"
+  name  = "${var.environment}_${var.queue_names[count.index]}"
+}
+
+# Create an IAM policy and attach it to the environment role
+# and give permissions to list queues and manage messages
+resource "aws_iam_role_policy" "sqs_manage_policy" {
+  name = "${var.environment}_sqs_manage_policy"
+  role = "${var.role}"
+
+  policy = <<EOF
+{
+    "Version": "2012-10-17",
+    "Statement": [
+        {
+            "Sid": "VisualEditor0",
+            "Effect": "Allow",
+            "Action": [
+                "sqs:DeleteMessage",
+                "sqs:SendMessage",
+                "sqs:ReceiveMessage",
+                "sqs:GetQueueAttributes",
+                "sqs:GetQueueUrl"
+            ],
+            "Resource": [
+                "arn:aws:sqs:${var.region}:${var.account_number}:${var.environment}_*"
```

Comment:
ok I hate this. Why can't you use the list of queues that come from aws_sqs_queue.qa_reports_queue? Having to pass in region and account_number is smelly.

Reply:
I hate it as much as you do! I just couldn't figure it out. Here's what I attempted:

1st:
```
"Resource": ["${aws_sqs_queue.qa_reports_queue.*.arn}"]
```

2nd:
```
"Resource": "${aws_sqs_queue.qa_reports_queue.*.arn}"
```

3rd:
```
locals {
  arns = ["${aws_sqs_queue.qa_reports_queue.*.arn}"]
}
...
"Resource": "${local.arns}"
```

all attempts above resulted in the error below:
```
* module.sqs.aws_iam_role_policy.sqs_manage_policy: 1 error(s) occurred:
* module.sqs.aws_iam_role_policy.sqs_manage_policy: At column 1, line 1: output of an HIL expression must be a string, or a single list (argument 2 is TypeList)
```

After some googling, my last attempt was this stackoverflow answer: https://stackoverflow.com/a/45816492. It loads the policy straight out of a json file, instead of using here doc syntax. But that throws a syntax error and I didn't debug it to see what happened exactly.

I tried using terraform data source as well, but it fails because queues don't yet exit.

Any hints?

Reply:
I'm not sure, but perhaps join() can help? I think `aws_sqs_queue.qa_reports_queue` is a list, right?

```
"Resource": [ "${join(", ", aws_sqs_queue.qa_reports_queue)}" ]
```

Reply:
Awesome! I hadn't thought of `join`. After minor tweaking, it ended up working with `["${join("\",\"", aws_sqs_queue.qa_reports_queue.*.arn)}"]`